fluentd/README.md

```diff
@@ -64,12 +18,49 @@ file chunk, perform a limited set of operations to confirm message validity. Use
+### Directories
+Dependencies are organized according in several basic directories:
+* `lib` - Dependencies for which this repo is considered the single source of truth.  This is where we make changes that are permanent forks of other libraries or libraries that do not have any other upstream
+* `vendored_gem_src` - Dependencies that are copied in from upstream sources based on what is defined in the `Gemfile`.  Changes should never be made to content in this directory as it is likely to be clobbered on updates
 
-Contents of files stored in the directory where `LOGGING_FILE_PATH` is, are printed out. Starting with the oldest log. Using `-f` option you can follow what is being written into the logs.
-The default path where the logs are is `/var/log/fluentd/`, default name of the log file is `fluentd.log`.
 
-## Changes
+### Modifying a gem Dependency
+Modifying or updating any of the gems requires a local ruby installation.  The version of ruby should match that defined in the `Dockerfile`.  It is highly recommended to setup the environment using something like [rvm](https://rvm.io/) which makes it possible to install multiple versions of ruby and define sets of gems.  The following instructions assume the use of rvm.
+#### Install ruby
+```
+$ rvm install ruby-2.5
+$ rvm use ruby-2.5
+```
+#### Install Vendored gems
+Gems required by fluentd are maintained in the [Gemfile](https://bundler.io/gemfile.html) which is the list of dependencies.  The repository also includes the `Gemfile.lock` which is the explicit list of the versions of the dependencies being used including the transitive ones.
 
-* 2018-Apr-20 - `merge_json_log` - Introduce variable for deprecating config option. [bug 1569825](https://bugzilla.redhat.com/show_bug.cgi?id=1569825)
+```
+$ rvm gemset create fluentd
+$ rvm gemset use fluentd
+$ gem install bundler -v 1.17.3
+$ bundle install
+```
+#### Updating Vendored Gems
+* Edit the Gemfile as needed
+* Run the update script from the repo root dir
+* Commit changes and rebuild image
+```
+$ ./hack/update-fluentd-vendor-gems.sh
+```
+**Note:** The hack script accepts the following env vars:
+* `BUNDLE_TASK`: 'install'(default) or 'update' to modify what bundler is doing.  'update' may be used if the intent is to update all the gems to the latest.  This will also update the `Gemfile.lock` which should additionally be committed.
+* `CLOBBER_VENDOR`: This removes everything in `vendored_gem_src` for cases where the intent is to clean up the vendored gems to remove unused dependencies
+
+#### Make lib Changes
+* May need to install additional dependencies from `lib/<gemdir>` using `bundle install`
+* Make any needed changes to the files in lib
+* Test change `bundle exec rake test`
+* Commit change and rebuild image
+
+### Building the Container Image
+Build the container image from the fluentd directory like:
+```
+$ podman build -t fluentd:test .
```

Comment:
@jcantrill it would be helpful to explain that you need access to registry.svc.ci.openshift.org inorder to build the image (i.e. make sure you have the needed credentials under ~/.docker/config.json)

Reply:
@jcantrill 

I used docker instead of podman (maybe you can add that as option in the text above)

Can you add something like this:

### Test Container Image with custom fluentd configuration

run the following command to mount custom local fluentd configuration against the container and start fluentd executable:
```
docker run -v $(pwd)/fluent-test-tail.conf:/etc/fluent/fluent-test-tail.conf -it fluentd:test fluentd -v -c /etc/fluent/fluent-test-tail.conf
```

Reply:
> @jcantrill it would be helpful to explain that you need access to registry.svc.ci.openshift.org in order to build the image (i.e. make sure you have the needed credentials under ~/.docker/config.json)

I will add a note but making the repo generally buildable outside of being an openshift dev is a longer term goal.  I don't believe this is the place to be defining larger needs around "how to hack logging"

Reply:
> @jcantrill
> 
> I used docker instead of podman (maybe you can add that as option in the text above)
> 
> (optionally) docker build -t fluentd:test .
> 
> Can you add something like this:
> 
> ### Test Container Image with custom fluentd configuration
> run the following command to mount custom local fluentd configuration (e.g ~/fluent-test-tail.conf) against the container and start fluentd executable:
> 
> ```
> docker run -v $(pwd)/fluent-test-tail.conf:/etc/fluent/fluent-test-tail.conf -it fluentd:test fluentd -v -c /etc/fluent/fluent-test-tail.conf
> ```
> 
What am i testing here? Why would I do this?

> @jcantrill Also if you can add this section to explain how to push the custom image back to OCP image registry::

This is part of generally hacking on logging, unrelated to fluentd. I would prefer to leave any of this info to any doc about "how do I interact and develop against OCP"
>

Reply:
/lgtm
I see there are outstanding requests, but bhe changes as they stand are an improvement. We can make additional changes later.